maxflowを作成

komajun365 opened the pull request.

komajun365 committed:
maxflowを作成

komajun365 commented:
ALPC提出結果
・D -  Maxflow：210 ms
https://atcoder.jp/contests/practice2/submissions/18716560

komajun365 commented:
関数min_cutのdocstring書き忘れてました。WIPに戻します。

komajun365 commented:
関数min_cutのdocstringを追加しました。
WIPを外します。

komajun365 committed:
Merge branch 'master' into feature/maxflow

fujitahub commented:
maxflow.pyの145, 146行目
`len(self._g[from_])`を使った後に`from_id`に同じ値を格納しているので、順番を入れ替えた方が良いのでは。
今：
```
self._pos.append([from_, len(self._g[from_])])
from_id = len(self._g[from_])
```

変更後：
```
from_id = len(self._g[from_])
self._pos.append([from_, from_id])
```

fujitahub commented:
maxflow.pyの163 - 165行目
内包表記で書いた方が良いのでは。
`result = [self.get_edge(i) for i in range(m)]`

fujitahub commented:
maxflow.py
`while`, `if`の書き方がpython likeじゃないです。
例えば
181行目 `while(que):` -> `while que:`
188行目 `if(to == t):` -> `if to == t:`

komajun365 commented:
> maxflow.pyの163 - 165行目
> 内包表記で書いた方が良いのでは。
> `result = [self.get_edge(i) for i in range(m)]`

内包表記とfor文の使い分け、
正直言ってこだわりがないので指摘の方に変更しましたが、
規約とかでどっちを書くべきみたいな推奨ってあるんでしょうか？

komajun365 committed:
レビュー指摘を受けて修正
・maxflow.pyの145, 146行目→変数を使いまわせる形に処理順番を変更
・maxflow.pyの163行目→内包表記に変更
・if,whie分の書き方をpython likeに変更（括弧を極力使わない）
・test_maxflowの無駄な行を一部削除

fujitahub commented:
内包表記の方が実行速度が速いため、内包表記で書くことができるときは内包表記で書く方が良いと認識しています。
